[RISC-V] coreclr-jit directory

clamp03 opened the pull request:
- Successfully cross-build for RISC-V.
- Run A simple application "helloworld"
- Fail a test in clr.paltest

clamp03 committed:
[RISCV-V] coreclr-jit directory

- Successfully cross-build for RISC-V.
- Run A simple application "helloworld"
- Fail a test in clr.paltest

clamp03 committed:
Fix in jit/CMakeLists.txt

BruceForstall commented:
@dotnet/jit-contrib

clamp03 committed:
Tidy up by jit-format

jakobbotsch commented on src/coreclr/jit/codegenlinear.cpp:
```diff
@@ -2059,9 +2063,15 @@ void CodeGen::genSpillLocal(unsigned varNum, var_types type, GenTreeLclVar* lclN
-        // Ensure that lclVar stores are typed correctly.
+// Store local variable to its home location.
+// Ensure that lclVar stores are typed correctly.
+#ifdef TARGET_RISCV64
+        assert(!compiler->isSIMDTypeLocalAligned(varNum));
+        GetEmitter()->emitIns_S_R(ins_Store(type, compiler->isSIMDTypeLocalAligned(varNum)), emitTypeSize(type), regNum,
+                                  REG_NA,
+#else
         GetEmitter()->emitIns_S_R(ins_Store(type, compiler->isSIMDTypeLocalAligned(varNum)), emitTypeSize(type), regNum,
+#endif
```
Can you add a new RISC-V only function for when the temporary register is necessary, to be used in the places that do not pass `REG_NA`? That way we can avoid adding all these ifdefs that just pass `REG_NA` anyway in shared code.

clamp03 replied:
Fixed! Thank you.

clamp03 committed:
Add a emitIns_S_R_R to handling more temporary reg

clamp03 committed:
Fix jit code format

clamp03 commented:
@dotnet/jit-contrib  Could you review this PR? Actually I don't know why tests failed or cancelled.
If you leave any comment, I will update as possible as I can.
Thank you.

BruceForstall commented:
@clamp03 As you know, this is a very large change and will take a significant amount of time and effort to review. We are discussing internally how to approach it, but no decisions have been made yet.

clamp03 commented:
@BruceForstall Thank you for the comment. I can wait. :)

clamp03 committed:
Fix

clamp03 committed:
[JIT] Fix test

jakobbotsch commented on src/coreclr/jit/codegencommon.cpp:
```diff
@@ -4191,6 +4198,34 @@ void CodeGen::genEnregisterIncomingStackArgs()
+            bool FPbased;
+            int  base = compiler->lvaFrameAddress(varNum, &FPbased);
+
+            if (emitter::isValidSimm12(base))
+            {
+                GetEmitter()->emitIns_R_S(ins_Load(regType), emitTypeSize(regType), regNum, varNum, 0);
+            }
+            else
+            {
+                if (tmp_reg == REG_NA)
+                {
+                    regNumber reg2 = FPbased ? REG_FPBASE : REG_SPBASE;
+                    tmp_offset     = base;
+                    tmp_reg        = REG_RA; // TODO CHECK R21 => RA
+
+                    GetEmitter()->emitIns_I_la(EA_PTRSIZE, REG_RA, base);                   // TODO CHECK R21 => RA
+                    GetEmitter()->emitIns_R_R_R(INS_add, EA_PTRSIZE, REG_RA, REG_RA, reg2); // TODO CHECK R21 => RA
+                    GetEmitter()->emitIns_R_S(ins_Load(regType), emitTypeSize(regType), regNum, varNum, -8);
+                }
+                else
+                {
+                    int baseOffset = -(base - tmp_offset) - 8;
+                    GetEmitter()->emitIns_R_S(ins_Load(regType), emitTypeSize(regType), regNum, varNum, baseOffset);
+                }
+            }
```
Why is this special casing on RISCV (and LA64, for that matter) necessary? It looks strange to me -- if RISCV cannot handle large displacements, then how does this handle the situation given that `tmp_offset`/`tmp_reg` is only updated once?

Other platforms use `CodeGen::rsGetRsvdReg` to handle reserving a temporary register to be used in situations of unencodable offsets. Can RISCV/LA64 be put on the same plan?

clamp03 replied:
Thank you. Removed codes and replaced temp reg use with rsGetRsvdReg.

jakobbotsch reviewed (commented):
First quick pass.

@clamp03 can you please avoid force-pushing when you address feedback? It makes it very hard for reviewers to see what changed.

SingleAccretion commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -17090,7 +17132,8 @@ const GenTreeLclVarCommon* GenTree::IsLocalAddrExpr() const
 GenTreeLclVarCommon* GenTree::IsImplicitByrefParameterValuePreMorph(Compiler* compiler)
 {
-#if FEATURE_IMPLICIT_BYREFS && !defined(TARGET_LOONGARCH64) // TODO-LOONGARCH64-CQ: enable this.
+#if FEATURE_IMPLICIT_BYREFS && !defined(TARGET_LOONGARCH64) &&                                                         \
+    !defined(TARGET_RISCV64) // TODO-LOONGARCH64-CQ & TODO-RISCV64: enable this.
```
Though I did not write the code, I believe this was simply missed in the LA64 bringup (LA64 predates `FEATURE_IMPLICIT_BYREFS`, so the places now under it were doing individual target `#ifdef`s).

I do not think there is a reason to pessimize RISCV here.

shushanhf replied:
This is not written by me.
It seems pushed within this
```
commit 5fda6fd23fa1d2732ca3418d5b6829dfd1e61a19
Author: Jakob Botsch Nielsen <Jakob.botsch.nielsen@gmail.com>
Date:   Mon Jan 23 17:43:26 2023 +0100

    JIT: Allow forwarding field accesses off of implicit byrefs (#80852)
    
    The JIT currently allows forwarding implicit byrefs at their last uses
    to calls, but only if the full implicit byref is used. This change
    allows the JIT to forward any such access off of an implicit byref 
    parameter.
```

jakobbotsch replied:
@shushanhf The original LA64 PR did not enable this function on LA64, even though it has implicit byrefs:
https://github.com/dotnet/runtime/blob/51afa067be9f0d00dac2c455ceb78e38eef1e239/src/coreclr/jit/gentree.cpp#L15872-L15884
(From the LA64 PR commit)

Since then @SingleAccretion and I refactored these functions, but the special case for LA64 remains due to the above. If this was just an oversight then can you please PR a change to enable this and `IsImplicitByrefParameterValuePostMorph` for LA64?

shushanhf replied:
OK, thanks
I will test it on LA64.

clamp03 replied:
Removed `!defined(TARGET_RISCV64)` in the PR.

clamp03 committed:
[JIT] Updated

clamp03 commented:
> First quick pass.
> 
> @clamp03 can you please avoid force-pushing when you address feedback? It makes it very hard for reviewers to see what changed.

Sorry. I did force-push again by mistake. I will not do again. Sorry again.
I updated by following your guide.

clamp03 commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -535,27 +535,27 @@ var_types Compiler::getPrimitiveTypeForStruct(unsigned structSize, CORINFO_CLASS
             break;
 
-#if !defined(TARGET_XARCH) || defined(UNIX_AMD64_ABI) || defined(TARGET_LOONGARCH64)
+#if !defined(TARGET_XARCH) || defined(UNIX_AMD64_ABI) || defined(TARGET_LOONGARCH64) || defined(TARGET_RISCV64)
```
Fixed!

clamp03 commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -3239,10 +3241,10 @@ void Compiler::compInitOptions(JitFlags* jitFlags)
 
-#ifdef TARGET_LOONGARCH64
+#if defined(TARGET_LOONGARCH64) || defined(TARGET_RISCV64)
     // Hot/cold splitting is not being tested on LoongArch64.
```
Removed `ifdef` and `endif`

clamp03 commented on src/coreclr/jit/emit.cpp:
```diff
@@ -3689,6 +3697,10 @@ const size_t hexEncodingSize = 11;
 const size_t basicIndent     = 12;
 const size_t hexEncodingSize = 19;
+#elif defined(TARGET_RISCV64)
+// TODO RISCV64
```
Removed

clamp03 commented on src/coreclr/jit/emit.cpp:
```diff
@@ -4068,6 +4080,7 @@ void emitter::emitDispIG(insGroup* ig, bool displayFunc, bool displayInstruction
 
         printf("\n");
 
+#if !defined(TARGET_RISCV64)
```
I put a TODO-RISCV64-Bug

jakobbotsch replied:
Does this mean `DOTNET_JitDisasm` does not work for RISC-V?

jakobbotsch replied:
Ping @clamp03 on the above question.

If `DOTNET_JitDisasm` does not work yet, is it in the plans to work on it to make it work?

clamp03 replied:
Actually, we don't any plan for `DOTNET_JitDisasm`. Currently, I and my colleagues are focusing on fixing bugs in coreclr tests using `DOTNET_JitDump`, `GDB`, and others. Using the tools, we can improve a pass rate of `CodeGenBringUp` tests from around 40% to 90%. However, if you believe that `DOTNET_JitDisasm` is a significant feature, we can prioritize it accordingly. And we can implement internally and then push a new PR for `DOTNET_JitDisasm` after this patch is merged.
Thank you.

jakobbotsch replied:
Thanks. So if I understand correctly, you are using GDB to view the code produced by the JIT?

We are ok with delaying disassembly support to a future PR, but we do think it is important for all our backends to support outputting the disassembly as part of JitDisasm and JitDump. In particular to allow JIT developers and other contributors to more easily pick up and work on RISC-V themselves. However, I will leave the prioritization of it up to you.

On that same topic it would be great if you at some point could open a "road map" issue to track some of these missing bits for RISC-V and how you are prioritizing them.

clamp03 replied:
`DOTNET_JitDump` is working although it is not perfectly.
We will open an issue for "road map" and "prioritization" after discussion.
Thank you!

clamp03 commented on src/coreclr/jit/emit.cpp:
```diff
@@ -6257,8 +6271,8 @@ emitter::instrDescAlign* emitter::emitAlignInNextIG(instrDescAlign* alignInstr)
-#ifdef TARGET_LOONGARCH64
-    // TODO-LoongArch64: support idDebugOnlyInfo.
+#if defined(TARGET_LOONGARCH64) || defined(TARGET_RISCV64)
+    // TODO LoongArch64 / RISCV64: support idDebugOnlyInfo.
```
Seperated `TODO`s.

clamp03 commented on src/coreclr/jit/emit.h:
```diff
@@ -778,6 +790,13 @@ class emitter
 #endif
 
+#ifdef TARGET_RISCV64
+        // TODO RISCV64
```
Removed

clamp03 commented on src/coreclr/jit/emit.h:
```diff
@@ -2961,7 +3059,9 @@ class emitter
 
 inline void emitter::instrDesc::checkSizes()
 {
+#ifndef TARGET_RISCV64 // TODO RISCV64
```
Removed

clamp03 commented on src/coreclr/jit/emitinl.h:
```diff
@@ -362,6 +362,42 @@ inline ssize_t emitter::emitGetInsAmdAny(instrDesc* id)
+        encodeMask |= 0x20;
+
+    id->idReg2((regNumber)encodeMask); // Save in idReg2
+
```
Please let me know when a PR for other archs is ready.

clamp03 commented on src/coreclr/jit/emitjmps.h:
```diff
@@ -53,6 +53,12 @@ JMP_SMALL(jmp   , jmp   , b      )
 JMP_SMALL(ne    , eq    , bne    )  // NE
 
+#elif defined(TARGET_RISCV64)
+// TODO RISCV64: adding other condition branches
```
Removed

clamp03 commented on src/coreclr/jit/codegencommon.cpp:
```diff
@@ -1592,6 +1592,9 @@ void CodeGen::genCheckOverflow(GenTree* tree)
             }
         }
+#elif defined(TARGET_RISCV64)
+        _ASSERTE(!"TODO RISCV64");
```
Removed

clamp03 commented on src/coreclr/jit/instr.cpp:
```diff
@@ -913,10 +920,12 @@ void CodeGen::inst_RV_TT(instruction ins, emitAttr size, regNumber op1Reg, GenTr
             {
                 emit->emitIns_Mov(ins, size, op1Reg, op2Desc.GetReg(), /* canSkip */ true);
             }
+#endif // !TARGET_RISCV64
```
Because IsMovInstruction is not implemented yet, I disabled the codes. Now I updated IsMovInstruction and removed `if defined`

clamp03 commented on src/coreclr/jit/instr.h:
```diff
@@ -55,6 +55,12 @@ enum instruction : unsigned
 
+    INS_lea,   // Not a real instruction. It is used for load the address of stack locals
+#elif defined(TARGET_RISCV64)
+    // TODO RISCV64
```
Removed

clamp03 commented on src/coreclr/jit/unwind.h:
```diff
@@ -42,7 +42,16 @@ const unsigned MAX_EPILOG_SIZE_BYTES = 200;
 #define UW_MAX_CODE_WORDS_COUNT 31
 #define UW_MAX_EPILOG_START_INDEX 0x3FFU
-#endif // TARGET_LOONGARCH64
+#elif defined(TARGET_RISCV64) // TODO RISCV64
```
Removed

clamp03 committed:
[JIT] JitFormat

clamp03 committed:
[JIT] Enable IsImplicitByrefParameterValuePostMorph

clamp03 committed:
[JIT] Update

- Use rsGetRsvdReg
- Remove unnecessary codes

clamp03 committed:
Merge remote-tracking branch 'upstream/main' into riscv_jit

echoix commented on src/coreclr/jit/instrsriscv64.h:
```diff
@@ -0,0 +1,247 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+/*****************************************************************************
+ *  LoongArch64 instructions for JIT compiler
```
Forgot to change the reference of LoongArch64 to riscv64.

clamp03 replied:
I implemented based on LoongArch, but I didn't update some codes and comments.
Thank you. I updated all.

echoix commented on src/coreclr/jit/lsrariscv64.cpp:
```diff
@@ -0,0 +1,1240 @@
+            index                = lea->Index();
+            cns                  = lea->Offset();
+
+            // On LOONGARCH we may need a single internal register
```
LoongArch reference in riscv-specific file.

echoix commented on src/coreclr/jit/lsrariscv64.cpp:
```diff
@@ -0,0 +1,1240 @@
+            // (when both conditions are true then we still only need a single internal register)
+            if ((index != nullptr) && (cns != 0))
+            {
+                // LOONGARCH does not support both Index and offset so we need an internal register
```
LoongArch reference in riscv-specific file. (Also)

echoix commented on src/coreclr/jit/lsrariscv64.cpp:
```diff
@@ -0,0 +1,1240 @@
+    {
+        // Don't assign the call target to any of the argument registers because
+        // we will use them to also pass floating point arguments as required
+        // by LOONGARCH64 ABI.
```
LoongArch reference in riscv-specific file. (Also)

echoix commented on src/coreclr/jit/lsrariscv64.cpp:
```diff
@@ -0,0 +1,1240 @@
+
+                if (isDstRegAddrAlignmentKnown && (size > FP_REGSIZE_BYTES))
+                {
+                    // TODO-LoongArch64: For larger block sizes CodeGen can choose to use 16-byte SIMD instructions.
```
TODO-riscv64 ?

echoix reviewed (commented):
I was curious about work on RISC-V, and noticed that there was some remains of the LoongArch64 from which new files were based on.

clamp03 committed:
[JIT] Replace to RISCV64

I implemented based on LOONGARCH.
Some comments in the codes are not updated yet.

jakobbotsch commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -3236,10 +3238,7 @@ void Compiler::compInitOptions(JitFlags* jitFlags)
 
-#ifdef TARGET_LOONGARCH64
-    // Hot/cold splitting is not being tested on LoongArch64.
     opts.compProcedureSplitting = false;
```
This seems to be overriding the value set above.

clamp03 replied:
Oops. Sorry!

jakobbotsch commented on src/coreclr/jit/ee_il_dll.cpp:
```diff
@@ -437,16 +437,14 @@ unsigned Compiler::eeGetArgSize(CORINFO_ARG_LIST_HANDLE list, CORINFO_SIG_INFO*
             // This struct is passed by reference using a single 'slot'
             return TARGET_POINTER_SIZE;
         }
-//  otherwise will we pass this struct by value in multiple registers
-#elif !defined(TARGET_ARM)
-        NYI("unknown target");
```
Can you keep this NYI and expand the earlier case instead to what fits RISC-V's calling convention?

clamp03 replied:
Thank you!

jakobbotsch commented on src/coreclr/jit/emitinl.h:
```diff
@@ -500,6 +534,38 @@ inline ssize_t emitter::emitGetInsAmdAny(instrDesc* id)
+    assert(REGNUM_BITS >= 6);
+    encodeMask = id->idReg1();
+
+    // if ((encodeMask & 0x01) != 0) // TODO CHECK S0 is FP
+    //    regmask |= RBM_S0;
```
Does this todo needs handling?

jakobbotsch commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -17079,7 +17120,8 @@ GenTreeLclVarCommon* GenTree::IsImplicitByrefParameterValuePreMorph(Compiler* co
 GenTreeLclVar* GenTree::IsImplicitByrefParameterValuePostMorph(Compiler* compiler, GenTree** addr)
 {
-#if FEATURE_IMPLICIT_BYREFS && !defined(TARGET_LOONGARCH64) // TODO-LOONGARCH64-CQ: enable this.
+#if FEATURE_IMPLICIT_BYREFS && !defined(TARGET_LOONGARCH64) &&                                                         \
+    !defined(TARGET_RISCV64) // TODO-LOONGARCH64-RISCV64-CQ: enable this.
```
As we discussed elsewhere we should be able to enable this for RISC-V.

clamp03 committed:
[JIT] Update emitter and fix bugs

clamp03 commented on src/coreclr/jit/instr.cpp:
```diff
@@ -439,7 +446,16 @@ void CodeGen::inst_RV_RV(instruction ins,
+#endif
+    {
+        GetEmitter()->emitIns_R_R(ins, size, reg1, reg2);
+    }
```
Thank you. I updated that it uses emitIns_Mov with ins_Copy. Thank you!

clamp03 committed:
[JIT] Fix a build error (remove GT_ADDR)

clamp03 committed:
[JIT] FIX BUILD ERROR

- Remove GTF_BLK_VOLATILE
- Replace GT_LCL_VAR_ADDR and GT_LCL_FLD_ADDR with GT_LCL_ADDR

clamp03 commented:
> @clamp03 As you know, this is a very large change and will take a significant amount of time and effort to review. We are discussing internally how to approach it, but no decisions have been made yet.

@BruceForstall 
Over a month passes from the first commit. Did you discuss and make any decision?
As you know, this PR is so huge and it has so many bugs and unimplemented codes.
However, I cannot make the runtime perfectly working on RISC-V with only this PR.
I need helps from MS, community and my colleagues. Actually, my colleagues are working internally and some bug patches which they made are waiting until the PRs are merged.
If you have any decisions or requirements for the PR, please let me know.
Thank you.

BruceForstall commented:
@clamp03 : As you know, @jakobbotsch has been looking at this. I believe he plans to take another look at the changes soon (perhaps this week), and hopefully will be able to merge the change soon.

jakobbotsch commented on src/coreclr/jit/codegenriscv64.cpp:
```diff
@@ -0,0 +1,7757 @@
+    if (EA_IS_RELOC(size))
+    {
+        assert(genIsValidIntReg(reg));
+        emit->emitIns_R_AI(INS_jalr, size, reg, imm); // for example: EA_PTR_DSP_RELOC
```
`jalr` here seems odd?

Note that you won't see `opts.compReloc` be true until you start trying to work on prejitting support, so that's probably why this works. Could just make this NYI for now.

clamp03 replied:
NYI added. Thank you!

jakobbotsch commented on src/coreclr/jit/codegenriscv64.cpp:
```diff
@@ -0,0 +1,7757 @@
+        {
+            if (attr == EA_4BYTE)
+            {
+                if ((treeNode->gtFlags & GTF_UNSIGNED) != 0)
```
I'm confused here, when is `GTF_UNSIGNED` ever set on a `GT_RETURN` node?

clamp03 replied:
Removed.

jakobbotsch commented on src/coreclr/jit/codegenriscv64.cpp:
```diff
@@ -0,0 +1,7757 @@
+                isImm = isImmed(treeNode);
+                if (isImm)
+                {
+                    if ((attr == EA_8BYTE) || (attr == EA_BYREF))
+                    {
+                        ins = INS_addi;
+                    }
+                    else
+                    {
+                        assert(attr == EA_4BYTE);
+                        ins = INS_addi;
+                    }
```
Is it intentional that both cases are `INS_addi`?

clamp03 replied:
Sorry. My mistake.

jakobbotsch commented on src/coreclr/jit/codegenriscv64.cpp:
```diff
@@ -0,0 +1,7757 @@
+                        imm = static_cast<int8_t>(imm);
+                    }
+                    break;
+                // case EA_2BYTE:
+                //    if (IsUnsigned)
+                //    {
+                //        imm = static_cast<uint16_t>(imm);
+                //    }
+                //    else
+                //    {
+                //        imm = static_cast<int16_t>(imm);
+                //    }
+                //    break;
```
It's not really clear to me why this handling exist at all, narrowing the immediate value here based on the non-actual type of the first operand does not really make sense.

clamp03 replied:
I removed all. Thank you.

jakobbotsch commented on src/coreclr/jit/codegenriscv64.cpp:
```diff
@@ -0,0 +1,7757 @@
+
+    instruction ins = INS_invalid;
+
+    assert(jcc->OperIs(GT_JCC) && (instruction)jcc->GetRegNum() == INS_bnez);
```
Why is `GetRegNum()` storing `instruction` here? Looks odd.

It's likely you should avoid producing `GT_JCC` at all for RISC-V. `GT_JCC` is based on using the CPU flags from the special flags register that exists on x64/x86/arm32/arm64. If this does not exist on `RISC-V` then you should just stick to `GT_JTRUE`/`GT_JCMP`.

I suggest that you introduce your own version of `LowerJTrue` inside lowerriscv64.cpp, similar to LA64. Then you can ensure that not `JCC` node is ever produced (this is the only place that creates `JCC`).

jakobbotsch commented on src/coreclr/jit/codegenriscv64.cpp:
```diff
@@ -0,0 +1,7757 @@
+        {
+            reg2        = rsGetRsvdReg();
+            ssize_t imm = src2->AsIntConCommon()->IconValue();
+            emit->emitIns_I_la(EA_PTRSIZE, rsGetRsvdReg(), imm);
```
What is the point of marking these immediates as contained if codegen is going to materialize them into a register anyway? Should `GT_BOUNDS_CHECK` just be removed from `IsContainableImmed` instead?

clamp03 replied:
I removed `GT_BOUNDS_CHECK` from `IsContainableImmed`.

jakobbotsch replied:
The special cased integer constant codegen in `genRangeCheck` can be removed now -- if they are not marked contained, then codegen will just handle them as usual and you can use their registers directly.

clamp03 replied:
Thank you! Fixed!!

jakobbotsch commented on src/coreclr/jit/codegenriscv64.cpp:
```diff
@@ -0,0 +1,7757 @@
+            regNumber allocReg = splitNode->GetRegNumByIdx(0);
+            var_types regType  = splitNode->GetRegType(0);
+
+            // For LA64's ABI, the split is only using the A7 and stack for passing arg.
```
```suggestion
            // For RISC-V's ABI, the split is only using the A7 and stack for passing arg.
```

clamp03 replied:
Thank you.

jakobbotsch commented on src/coreclr/jit/codegenriscv64.cpp:
```diff
@@ -0,0 +1,7757 @@
+                preservedAreaSize += REGSIZE_BYTES;
+            }
+
+            preservedAreaSize += 1; // bool for synchronized methods
```
```suggestion
            preservedAreaSize += TARGET_POINTER_SIZE;
```

This was changed a while ago.

jakobbotsch replied:
Maybe just add NYI for this entire block? There are references to x64 registers above.

clamp03 replied:
Added NYI for the block. Thank you!

clamp03 committed:
[JIT] Update by reviews

clamp03 committed:
[JIT] Update JCMP and fix related errors

clamp03 committed:
[JIT] Add NYI

jakobbotsch commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -24047,7 +24088,7 @@ void ReturnTypeDesc::InitializeStructReturnType(Compiler*                comp,
+#elif defined(TARGET_LOONGARCH64) || defined(TARGET_RISCV64)
             assert((structSize >= TARGET_POINTER_SIZE) && (structSize <= (2 * TARGET_POINTER_SIZE)));
 
             uint32_t floatFieldFlags = comp->info.compCompHnd->getLoongArch64PassStructInRegisterFlags(retClsHnd);
```
Seems a bit odd to call `getLoongArch64PassStructInRegisterFlags` here, but I guess the ABIs are close enough/equivalent for this to work out?

clamp03 replied:
Sorry. It is really odd even if these are very close. I created a new for RISCV64 and pushed a commit in both vm and jit PRs.
Thank you.

jakobbotsch replied:
This also requires updating the JIT-EE GUID and ThunkInput.txt. See https://github.com/dotnet/runtime/blob/main/docs/project/updating-jitinterface.md for more details.

jakobbotsch replied:
I don't think it is ok to leave it unimplemented on the VM side in this PR. Can you introduce a skeleton for this new JIT-EE API? If you need some more help then #78593 is an example PR that introduces a new JIT-EE API.

clamp03 replied:
I already added ThunkInput.txt and the other codes into https://github.com/dotnet/runtime/pull/82381. (*. I missed GUID update.)

If you think that all related codes should be in a PR, then how about reverting the last commit and make a new PR after this coreclr-jit and coreclr-vm (https://github.com/dotnet/runtime/pull/82381) PR are merged?
If not, I will update GUID in https://github.com/dotnet/runtime/pull/82381 and so on.
Thank you

jakobbotsch replied:
The problem is that we cannot merge this PR without coherent JIT-EE implementations/consumptions for all our tooling. I would suggest adding the helper (including the VM side and the rest of the boilerplate) as part of this PR.

clamp03 replied:
I moved coreclr-vm commit to this and updated GUID.
Thank you.

jakobbotsch commented on src/coreclr/jit/lowerriscv64.cpp:
```diff
@@ -0,0 +1,880 @@
+            case GT_XOR:
+                return emitter::isValidUimm11(immVal);
+            case GT_JCMP:
+                assert(((parentNode->gtFlags & GTF_JCMP_TST) == 0) ? (immVal == 0) : isPow2(immVal));
```
Does RISC-V have the equivalent of ARM64's `tbnz`/`tbz` instructions?

brucehoult replied:
Nope. In general, replace `TNBZ Rt, #imm, label` with  `SLLI tmp, Rt, 63-imm; BLTZ tmp, label` and TBZ with the same but BGEZ. 2 byte instruction plus 4 byte branch instruction.

For imm in 0..10 you could use ANDI with 1<<imm, then BEQZ or BNEZ, which is a 4 byte instruction then a 2 byte branch...  But it's the same size and performance anyway

clamp03 replied:
@BruceForstall Thank you so much.

BruceForstall replied:
@brucehoult , that is :)

@brucehoult -- looks like you have lots of RISC-V experience. Maybe we'll see you contribute to the .NET implementation?

brucehoult replied:
> @brucehoult -- looks like you have lots of RISC-V experience.

Since you asked :-) I worked on the CoreCLR to Arm port at Samsung R&D Moscow in 2016, bought one of the first HIFive1 RISC-V boards in Jan 2017, then in early 2018 left Samsung to join RISC-V pioneers SiFive in San Mateo where I worked on runtime, compiler, emulator stuff and also contributed to designing the B and V extensions. (and less so others). You'll find my name in the credits for those extensions, plus the unprivileged ISA manual. Also co-moderator of /r/riscv on Reddit since December 2019, in which time we've grown it from 2500 to 14000 members.

Today was my last day at a New Zealand web dev company I've been working at during COVID. Looking at moving back into the RISC-V world now it's starting to explode. Have VisionFive 2 :-)  Currently have ssh access to a 64 core 2.0 GHz OoO (similar to Cortex A72) SG2042 and doing a little vector 0.7.1 asm work on that for someone ... mmmm, tasty.

jakobbotsch replied:
Thanks @brucehoult! Good luck in the RISC-V world :)

@clamp03 If there is no direct equivalent of `tbnz`/`tbz` then RISC-V should not be setting this `GTF_JCMP_TST` flag. Indeed, I can see that it is never set by `LowerJTrue`, so this part of the asset here is just dead code.

clamp03 replied:
I removed the assert code in the last patch `https://github.com/dotnet/runtime/pull/82379/commits/607dd64ec3dbefa2115317fccc95a6a3df483a14`

jakobbotsch reviewed (approved):
LGTM with a couple of last questions! Thanks for the patience.

I will run some stress jobs just to ensure nothing else seems to be affected.

jakobbotsch commented:
/azp run runtime-coreclr jitstress, runtime-coreclr libraries-jitstress

clamp03 committed:
[JIT] Add getRISCV64PassStructInRegisterFlags

clamp03 committed:
[VM] Add getRISCV64PassStructInRegisterFlags

clamp03 committed:
[JIT] Remove constant handling in genRangeCheck

clamp03 committed:
[VM] Update JITEEVersionIdentifier

clamp03 committed:
Merge remote-tracking branch 'upstream/main' into riscv_jit

jakobbotsch committed:
Run jit-format

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into pr-82379

jakobbotsch committed:
Fix after merge

jakobbotsch commented:
@clamp03 Thanks again for the patience and for addressing all the feedback!

clamp03 commented:
@jakobbotsch Thank you so much. I am really appreciate for your reviews and hard works.
